sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java

```diff
@@ -596,28 +620,64 @@ public void remove(Collection<String> filenames) throws IOException {
     executeBatches(makeRemoveBatches(filenames));
   }
 
-  private StorageObject[] enqueueGetFileSize(final GcsPath path, BatchRequest batch)
+  private StorageObjectOrIOException[] enqueueGetFileSize(final GcsPath path, BatchRequest batch)
       throws IOException {
-    final StorageObject[] storageObject = new StorageObject[1];
+    final StorageObjectOrIOException[] ret = new StorageObjectOrIOException[1];
 
     Storage.Objects.Get getRequest = storageClient.objects()
         .get(path.getBucket(), path.getObject());
     getRequest.queue(batch, new JsonBatchCallback<StorageObject>() {
       @Override
       public void onSuccess(StorageObject response, HttpHeaders httpHeaders) throws IOException {
-        storageObject[0] = response;
+        ret[0] = StorageObjectOrIOException.builder()
+            .setStorageObject(response)
+            .build();
       }
 
       @Override
       public void onFailure(GoogleJsonError e, HttpHeaders httpHeaders) throws IOException {
+        IOException ioException;
         if (errorExtractor.itemNotFound(e)) {
-          throw new FileNotFoundException(path.toString());
+          ioException = new FileNotFoundException(path.toString());
         } else {
-          throw new IOException(String.format("Error trying to get %s: %s", path, e));
+          ioException = new IOException(String.format("Error trying to get %s: %s", path, e));
         }
+        ret[0] = StorageObjectOrIOException.builder()
+            .setIoException(ioException)
+            .build();
       }
     });
-    return storageObject;
+    return ret;
+  }
+
+  /**
+   * A class that holds either a {@link StorageObject} or an {@link IOException}.
+   */
+  @AutoValue
+  public abstract static class StorageObjectOrIOException {
+
+    /**
+     * Returns the {@link StorageObject}.
+     */
+    @Nullable
+    public abstract StorageObject storageObject();
+
+    /**
+     * Returns the {@link IOException}.
+     */
+    @Nullable
+    public abstract IOException ioException();
+
+    static Builder builder() {
```

Comment:
The way this is used you could just have two different static constructors, one which takes a (non-null) `StorageObject` and one which takes a (non-null) `IOException`